bootloader/uboot.go

```diff
@@ -60,18 +89,49 @@ func (u *uboot) dir() string {
 	if u.rootdir == "" {
 		panic("internal error: unset rootdir")
 	}
-	if u.noslashboot {
-		return u.rootdir
-	}
-	return filepath.Join(u.rootdir, "/boot/uboot")
+	return filepath.Join(u.rootdir, u.basedir)
 }
 
-func (u *uboot) InstallBootConfig(gadgetDir string, opts *Options) (bool, error) {
+func (u *uboot) InstallBootConfig(gadgetDir string, blOpts *Options) (bool, error) {
 	gadgetFile := filepath.Join(gadgetDir, u.Name()+".conf")
-	systemFile := u.ConfigFile()
-	if opts != nil && opts.Recovery {
-		systemFile = filepath.Join(u.rootdir, "uboot.env")
+	// if the gadget file is empty, then we don't install anything
+	// this is because there are some gadgets, namely the 20 pi gadget right
+	// now, that don't use a uboot.env to boot and instead use a boot.scr, and
+	// installing a uboot.env file of any form in the root directory will break
+	// the boot.scr, so for these setups we just don't install anything
+	// TODO:UC20: how can we do this better? maybe parse the file to get the
+	//            actual format?
+	st, err := os.Stat(gadgetFile)
+	if err != nil {
+		return false, err
+	}
+	if st.Size() == 0 {
+		// we have a uboot.conf, and hence a uboot bootloader in the gadget, so
+		// we don't copy anything from the gadget and instead just install our
+		// own boot.sel file
+		u.processBlOpts(blOpts)
+
+		err := os.MkdirAll(filepath.Dir(u.envFile()), 0755)
+		if err != nil {
+			return false, err
+		}
+		_, err = ubootenv.Create(u.envFile(), u.ubootEnvFmt, 0)
+		if err != nil {
+			return false, err
+		}
+
+		return true, nil
 	}
+
+	// InstallBootConfig gets called on a uboot that does not come from newUboot
+	// so we need to apply the defaults here
+	u.setDefaults()
+
+	if blOpts != nil && blOpts.Recovery {
```

Comment:
it seems here that Recovery means a different thing than in processBlOpts, something is either off or confusing

Reply:
well I wasn't sure what to do here, since effectively this would only happen if we used a uboot gadget with uc20 that _only_ uses uboot.env, and thus we need to install the version from the gadget. So in effect it is a different thing than in processBlOpts, but we don't really have a use case for it right now other than the unit test I wrote in makebootable_test.go in TestUbootMakeBootable20TraditionalUbootenv

Reply:
as discussed on IRC, I will drop this check and now if uboot.conf is non-empty and we have Recovery it will return a "not implemented" error